shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs

```diff
@@ -0,0 +1,273 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Cardano.Ledger.ShelleyMA.TxBody
+  ( TxBody (TxBody, STxBody),
+    TxId (..),
+    TxIn (..),
+    TxOut (..),
+    ChainData,
+    TxBody' (..),
+  )
+where
+
+-- This is the type family
+import Cardano.Binary (Annotator, FromCBOR (..), ToCBOR (..))
+import Cardano.Ledger.Core (Value)
+import Cardano.Ledger.Era (Crypto, Era)
+import Cardano.Ledger.ShelleyMA.Timelocks (ValidityInterval (..), decodeVI, encodeVI)
+import Control.DeepSeq (NFData)
+import Data.ByteString.Short (ShortByteString, fromShort, toShort)
+import Data.Coders
+  ( Decode (..),
+    Encode (..),
+    decode,
+    decodeStrictSeq,
+    encode,
+    (!>),
+    (<!),
+  )
+import Data.MemoBytes (Mem, MemoBytes (..), memoBytes)
+import Data.Sequence.Strict (StrictSeq)
+import Data.Set (Set)
+import Data.Typeable (Typeable)
+import Data.Word (Word64)
+import GHC.Generics (Generic)
+import GHC.Records
+import GHC.TypeLits (Symbol)
+import NoThunks.Class (NoThunks (..))
+import Shelley.Spec.Ledger.Address (Addr (..))
+import Shelley.Spec.Ledger.BaseTypes (StrictMaybe)
+import Shelley.Spec.Ledger.Coin (Coin (..))
+import Shelley.Spec.Ledger.Keys (Hash)
+import Shelley.Spec.Ledger.MetaData (MetaDataHash)
+import Shelley.Spec.Ledger.PParams (Update)
+import Shelley.Spec.Ledger.Serialization (encodeFoldable)
+import Shelley.Spec.Ledger.TxBody (DCert (..), Wdrl (..))
+
+-- =========================================================================================================
+-- What do we require of TxBody? That it lives on the Block Chain, so it must have these instances.
+
+type ChainData t = (Eq t, Ord t, Show t, FromCBOR t, ToCBOR t, NoThunks t, NFData t, Typeable t)
+
+--type Value era = AbstractValue.Value era
+
+-- ================================================
+-- We cannot use the TxId from Shelley.Spec.Ledger.TxBody, because TxId and TxBody
+-- are mutually recursive (through the second (phantom) type index of Hash).
+-- TxId -> TxBody -> TxIn -> TxId. If we use the
+-- TxId from Shelley.Spec.Ledger.TxBody we get unsatisfiable constraints that
```

Comment:
Jared also introduced `EraIndependentBlockBody` to get around this phantom type issue for the body hash, so an alternative would be to do the same here.

Reply:
I introduced a nice abstraction parameterisng TxId by the phantom type.
Every thing becomes much simpler. This is well documented in the code now.